File: src/problems_all/1491_average_salary_excluding_the_minimum_and_maximum_salary/1491_average_salary_excluding_the_minimum_and_maximum_salary_v1.go
/*
https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/

O(n)
Runtime: 0 ms
Memory: 1.9 MB
https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/submissions/874163609/
*/

package main

import (
	"fmt"
	"math"
)

func main() {
	salary := []int{48000, 59000, 99000, 13000, 78000, 45000, 31000, 17000, 39000, 37000, 93000, 77000, 33000, 28000, 4000, 54000, 67000, 6000, 1000, 11000}
	result := average(salary)
	fmt.Println(result)
}

func minMaxSum(arr []int) (int, int, int) {
	min := math.MaxInt64
	max := math.MinInt64
	sum := 0

	for _, i := range arr {
		if i > max {
			max = i
		}

		if i < min {
			min = i
		}

		sum += i
	}

	return min, max, sum
}

func average(salary []int) float64 {
	min, max, sum := minMaxSum(salary)
	avg := float64(sum-min-max) / float64(len(salary)-2)

	return avg
}

File: src/problems_all/1491_average_salary_excluding_the_minimum_and_maximum_salary/1491_average_salary_excluding_the_minimum_and_maximum_salary_v1.py
"""
https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/

O(n)
Runtime: 41 ms
Memory: 13.8 MB
https://leetcode.com/problems/average-salary-excluding-the-minimum-and-maximum-salary/submissions/865241087/
"""

from typing import List


class Solution:
    def average(self, salary: List[int]) -> float:
        highest = max(salary)
        lowest = min(salary)
        average = (sum(salary) - highest - lowest) / (len(salary) - 2)
        return average


def main():
    salary = [4000, 3000, 1000, 2000]
    avg = Solution().average(salary)
    print(avg)


if __name__ == '__main__':
    main()
